cli/doc/man/man1/virgil.1
.Dd April 22, 2015
.Dt VIRGIL 1
.Os Darwin
.Sh NAME
.Nm virgil
.Nd command line interface to the Virgil library.
.Sh SYNOPSIS
.Nm
.Ar command
.Op Ar command_opts
.Op Ar command_args
.Sh DESCRIPTION
The
.Nm
program is a command line tool for using the various cryptography functions of the Virgil library. It can be used to:
.Bl -bullet -offset "---"
.It
genarate private key
.It
create certificate
.It
encrypt data
.It
decrypt data
.It
sign data
.It
verify data
.El
.Sh AVAILABLE COMMANDS
.Bl -tag -width "key2cert"
.It \fBkeygen\fP
Generate private key with given parameters.
.It \fBkey2pub\fP
Extract public key from the private key.
.It \fBpub2cert\fP
Create certificate from the public key and identifiers.
.It \fBcertinfo\fP
Output detailed information about given certificate.
.It \fBencrypt\fP
Encrypt data for given recipients which can be defined by certificates and by passwords.
.It \fBdecrypt\fP
Decrypt data for given recipient which can be defined by certificate or by password.
.It \fBsign\fP
Sign data with private key.
.It \fBverify\fP
Verify data with certificate.
.El
.Sh SEE ALSO
.Xr virgilkeygen 1 ,
.Xr virgilkey2cert 1 ,
.Xr virgilencrypt 1 ,
.Xr virgildecrypt 1 ,
.Xr virgilsign 1 ,
.Xr virgilverify 1


ChangeLog
virgil ChangeLog (Sorted per date)

= Version 0.8.0 released 2015-04-16

Bugfix:
  * Lib: Store algorithm parameters in ASN.1 structure when marshal cipher with elliptic-curve key.
  * Programs: Add omitted parameter 'certificate id'.

Features:
  * AS3: Add class VirgilBase64.
  * Ruby. Add wrapper for the ruby.
  * Python. Add wrapper for the python.
  * Emscripten: Add emscripten(asm.js) wrapper.

Improvements:
  * Lib: Encapsulate fields of the class VirgilCipherBase.
  * Lib: Hide usage of classes: VirgilAsn1Reader and VirgilAsn1Writer.
  * CMake: Remove redundant installation of library headers.
  * CMake: Add optional flag LOW_LEVEL_API - default value is OFF.
  * CMake: Remove unused variable - SIMULATOR.

Changes:
  * AS3: Move classes 'VirgilCustomParams' and 'VirgilContentInfo' to the package 'com.virgilsecurity.crypto'.
  * Lib: Add member type 'value_type' to the VirgilIdProvider.
  * Lib: Move version configuration from header file to the implementation file.
  * CMake: Change library installation policy, now target folder depends on the machine architecture.

= Version 0.7.0 released 2015-03-31
Bugfix:
  * CMake: Fix CMAKE_OSX_ARCHITECTURES variable default definition.

Features:
  * C#: Add mapping C++ VirgilByteArray to the C# byte[].
  * C#: Add runtime selection of the library architecture: x86 or x86_64.
  * AS3: Add wrapper for class VirgilHash.

Improvements:
  * C#: Set assembly version to the library current version.
  * AS3: Move utility headers to the appropriate place.
  * SWIG: Simplify including type mappings for VirgilByteArray.
  * SWIG: Move language specific typemaps to separate files.

Changes:
  * All: Rename method VirgilChunkCipher::finalize() to
         VirgilChunkCipher::finish() due to name confilcts on some platforms.
  * SWIG C#: Make first letter of all functions and properties uppercase.
  * CMake C#. Rename .NET assembly of the shared library from 'virgil.assembly' to 'Virgil.Net'
  * CMake: Selection of install directory for C# now depends on variables:
           CMAKE_SYSTEM_NAME, CMAKE_OSX_ARCHITECTURES, COMPILER_ARCH_32,
           COMPILER_ARCH_64.
  * SWIG: Ignore 'operator=' and 'operator=='.

= Version 0.6.0 released 2015-03-23
Bugfix:
  * CMake: Fix build options for jsoncpp library.
  * CMake: Define max java heap size for AS3 compilation process to prevent
           build fail.
  * AS3: Add try-catch section to the 'VirgilChunkCipher::startEncryption()'
         method wrapper.

Improvements:
  * CMake: Add version to the AS3 library file name.
  * CMake: Display full AS3 library name (including version) when build.
  * CMake: Prevent copy C++ headers to the AS3 installation folder.
  * CMake: Make AS3 library build only if needed.
  * CMake: Replace variable AS3_MODULE_FILE_NAME with synonym WRAPPER_SWC.
  * AS3: Change gradle configuration file to work with gradle version 2.3
       and gradlefx version 1.1.0.
  * CMake: Use UPDATE command instead of PATCH command in the
           ExternalProject_Add.
  * Lib: Add SHA-384 algorithm.
  * CMake: Add AS3 wrapper dependency on the library 'virgil'.

Changes:
  * AS3 Demo: Change demo app to be conformed with library version 0.5.0.
  * AS3 Demo: Embed content info to the encrypted file.
  * PHP: Remove 'iconv' and 'mb_detect_encoding' dependencies from unit tests.
  * All: Switch from 'polarssl' main branch to the 'mbed TLS' fork made by
         Virgil Security.
  * CMake: Refactor 'mbed TLS' patch. Now its configure files: 'config.h'
           and 'check_config.h' only.
  * Lib: PBE: use external functions.
  * Lib: Change SHA-256 to SHA-384 as default signer hash algorithm.
  * Lib: Rename VirgilByteArray utility functions to short versions.
  * Lib: Function 'getContentInfo()' become available immediately after
         function 'startEncryption()' call in the class VirgilChunkCipher.
  * AS3: Build AS3 unit tests in the strict mode.

= Version 0.5.0 released 2015-03-12
Features:
  * AS3: Implement Action Script 3 wrapper for virgil library version 0.4.0.

Improvements:
  * AS3: Speed and size optimization by using flascc Link Time Optimizer (LTO).
  * AS3: Make all tests configurable.
  * AS3: Configure test VirgilVersionTest with CMAKE variables.

Changes:
  * CMake: Apply platform specific compiler options only for native platforms
           build.
  * Lib: Changed marshalling / demarshalling process for tickets.
         Now 'class_name' field is handled by the base class.
  * Lib: Rename VirgilCipherBase::customParameters() to
         VirgilCipherBase::customParams().
  * Lib: Renaming VirgilVersion class methods.
  * Lib: Remove macroses: VIRGIL_BYTE_ARRAY_FROM_C_STRING,
         VIRGIL_BYTE_ARRAY_FROM_STD_STRING, VIRGIL_BYTE_ARRAY_TO_STD_STRING,
         VIRGIL_BYTE_ARRAY_ZEROIZE. Use inline functions analogue instead.
  * Lib: Remove 'jsoncpp' library dependency from the public API.


cli/src/virgil.cxx
/**
 * Copyright (C) 2014 Virgil Security Inc.
 *
 * Lead Maintainer: Virgil Security Inc. <support@virgilsecurity.com>
 *
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are
 * met:
 *
 *     (1) Redistributions of source code must retain the above copyright
 *     notice, this list of conditions and the following disclaimer.
 *
 *     (2) Redistributions in binary form must reproduce the above copyright
 *     notice, this list of conditions and the following disclaimer in
 *     the documentation and/or other materials provided with the
 *     distribution.
 *
 *     (3) Neither the name of the copyright holder nor the names of its
 *     contributors may be used to endorse or promote products derived from
 *     this software without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE AUTHOR ''AS IS'' AND ANY EXPRESS OR
 * IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
 * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
 * DISCLAIMED. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY DIRECT,
 * INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
 * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
 * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT,
 * STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING
 * IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
 * POSSIBILITY OF SUCH DAMAGE.
 */

#ifndef SPLIT_CLI

#include <iostream>
#include <string>
#include <map>

#include "utils.h"

typedef int (*main_func)(int argc, char **argv);

int keygen_main(int argc, char **argv);
int key2pub_main(int argc, char **argv);
int pub2cert_main(int argc, char **argv);
int certinfo_main(int argc, char **argv);
int encrypt_main(int argc, char **argv);
int decrypt_main(int argc, char **argv);
int sign_main(int argc, char **argv);
int verify_main(int argc, char **argv);

static void print_usage(std::ostream& out, const char *programName) {
    out << std::endl << "USAGE:" << std::endl;
    out << "    " << programName << " command [command_opts] [command_args]" << std::endl;
    out << std::endl << "DESCRIPTION:" << std::endl;
    out << "    " << "The virgil program is a command line tool for using the various "
            "cryptography functions of the Virgil library." << std::endl;
    out << std::endl << "AVAILABLE COMMANDS:" << std::endl;
    out << "    " << "keygen    " <<
            "Generate private key with a given parameters." << std::endl;
    out << "    " << "key2pub  " <<
            "Extract public key from the private key." << std::endl;
    out << "    " << "pub2cert  " <<
            "Create certificate from the public key and identifiers." << std::endl;
    out << "    " << "certinfo  " <<
            "Output detailed information about given certificate." << std::endl;
    out << "    " << "encrypt   " <<
            "Encrypt data for given recipients which can be defined by certificates and by passwords." << std::endl;
    out << "    " << "decrypt   " <<
            "Decrypt data for given recipient which can be defined by certificate or by password." << std::endl;
    out << "    " << "sign      " <<
            "Sign data with private key." << std::endl;
    out << "    " << "verify    " <<
            "Verify data with certificate." << std::endl;
}

static void print_version(std::ostream& out, const char *programName) {
    out << programName << "  " << "version: "<< virgil::cli::version() << std::endl;
}

int main(int argc, char **argv) {
    std::map<std::string, main_func> commandsMap;
    commandsMap["keygen"] = &keygen_main;
    commandsMap["key2pub"] = &key2pub_main;
    commandsMap["pub2cert"] = &pub2cert_main;
    commandsMap["certinfo"] = &certinfo_main;
    commandsMap["encrypt"] = &encrypt_main;
    commandsMap["decrypt"] = &decrypt_main;
    commandsMap["sign"] = &sign_main;
    commandsMap["verify"] = &verify_main;
    // Parse arguments.
    if (argc < 2) {
        std::cerr << "Error: " << " Required argument missing: " << "command" << std::endl;
        print_usage(std::cerr, argv[0]);
        return EXIT_FAILURE;
    }

    std::string firstArg(argv[1]);
    if (firstArg == "-h" || firstArg == "--help") {
        print_usage(std::cout, argv[0]);
        return EXIT_SUCCESS;
    } else if (firstArg == "--version") {
        print_version(std::cout, argv[0]);
        return EXIT_SUCCESS;
    }

    std::map<std::string, main_func>::const_iterator command = commandsMap.find(firstArg);
    if (command != commandsMap.end()) {
        command->second(argc - 1, argv + 1);
    } else {
        std::cerr << "Error: " << "command '" << firstArg << "' not found" << std::endl;
        return EXIT_FAILURE;
    }

    return EXIT_SUCCESS;
}

#endif /* SPLIT_CLI */
